### PR TITLE
slave_node: how to set up a bare metal jenkins slave node

### PR DESCRIPTION
**Pull request by Pennyzct**
Last week, I have added two new ARM CI nodes in the Kata Community.
Here, based on the instructions via emails from @grahamwhaley and @chavafg and my own experience, I'd like to write a tutorial on How to set up a bare metal Jenkins Slave node. ;)

Fixes: #166

Signed-off-by: Penny Zheng <penny.zheng@arm.com>

### CODE REVIEW

File: Jenkins_setup.md [hunk 1]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
 ![Jenkins Build dialog](pictures/jenkins_post_build.png)
 
+## Setting up a bare metal Jenkins Slave Node
+
+### Setting up a jenkins user
+
+Generally, Jenkins Master needs to log into the slave node as `jenkins` user to run the
+all Jenkins job.
+So you need to create a new user `jenkins` on your slave node and add this new user to
+the `sudo` group, which grants it `sudo` priviledges.
```
Comment (jodh-intel) [inline Jenkins_setup.md]: Nit: typo - "privileges".

File: Jenkins_setup.md [hunk 2]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+
+```
+$ sudo adduser jenkins
+$ sudo usermod -aG sudo jenkins
+```
+
+Not only just having sudo priviledges, the `jenkins` user also needs to be able to run `sudo`
```
Comment (jodh-intel) [inline Jenkins_setup.md]: Nits:

- consistency: `sudo`.
- typo: "privileges".

File: Jenkins_setup.md [hunk 3]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+without a password.
+Log into your slave node as root and open the `/etc/sudoers` file with `visudo` editor.
```
Comment (jodh-intel) [inline Jenkins_setup.md]: Could you put `root` in backticks?

File: Jenkins_setup.md [hunk 4]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+> **Note** You should never edit `/etc/sudoers` with a regular text editor, such as `Vim`
+> or `nano`, because they do not validate the syntax like the `visudo` editor.
+```
+sudo visudo
```
Comment (jodh-intel) [inline Jenkins_setup.md]: Can you add a `$ ` shell prefix? See https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md#code-blocks.

File: Jenkins_setup.md [hunk 5]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+```
+Add following line at the end of the `/etc/sudoers` file.
```
Comment (jodh-intel) [inline Jenkins_setup.md]: You could change these "manual instructions" into the magic:

```sh
$ echo "jenkins  ALL=(ALL) NOPASSWD:ALL" | (EDITOR="tee -a" visudo)
```

File: Jenkins_setup.md [hunk 6]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+```
+jenkins  ALL=(ALL) NOPASSWD:ALL
+```
+
+### Providing ssh private keys
+
+We need your host ssh private keys, which will lodge in the Jenkins master private key store.
```
Comment (jodh-intel) [inline Jenkins_setup.md]: - s/ssh/SSH/
- s/will lodge/will be stored/ ?

File: Jenkins_setup.md [hunk 7]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+The host machine should be the one in which you could remotely ssh into the slave CI Node
+as `jenkins` user.
+Generally, You could find the private ssh keys in `~/.ssh/id_rsa` in your host machine.
+
+> **Note** You could contact Graham Whaley ([`grahamwhaley`](https://github.com/grahamwhaley)) or
```
Comment (jodh-intel) [inline Jenkins_setup.md]: Although @grahamwhaley and @chavafg are probably the right people to contact, I'd rather we didn't "hard-code" them into a document ;) I'd add a link to https://github.com/kata-containers/community#join-us and suggest the reader contact the team on IRC or slack.

File: Jenkins_setup.md [hunk 8]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+> Salvador Fuentes ([`chavafg`](https://github.com/chavafg)) to upload your ssh private keys.
+
+### Pre-installled requirements
+
+You need to pre-install the following items in your bare metal Jenkins Slave CI node:
+
+- `JRE` for the Jenkins agent
```
Comment (jodh-intel) [inline Jenkins_setup.md]: I'd spell this out to make it clearer:

> - a Java Runtime Environment (JRE) for the Jenkins agent.

File: Jenkins_setup.md [hunk 9]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+- `golang`
+   a not-too-old version of go, lately the CI scripts will update it to the correct version.
```
Comment (jodh-intel) [inline Jenkins_setup.md]: A know this is awkward to explain but this is rather vague - how old is "too old"? Any thoughts on this @chavafg, @grahamwhaley?

File: Jenkins_setup.md [hunk 10]
```diff
@@ -142,6 +146,52 @@ The teardown script is entered into the Jenkins Post-Build Actions dialog like:
...
+- `git`
```
Comment (jodh-intel) [inline Jenkins_setup.md]: Nit: You could sort this list alphabetically :smile:

File: Jenkins_setup.md (hunk 6 above)
Comment (grahamwhaley) [reply Jenkins_setup.md]: I'd probably rewrite this to drop the 'we' etc....

---

You will need to set up an SSH key-pair for the `jenkins` user on your Jenkins slave, and provide the SSH private key to the Kata Containers CI administrators to be stored in the Jenkins master secure key store.

---

OK?

File: Jenkins_setup.md (hunk 7 above)
Comment (grahamwhaley) [reply Jenkins_setup.md]: ack - better not to hard code people, as over time their roles can change etc.

File: Jenkins_setup.md (hunk 9 above)
Comment (grahamwhaley) [reply Jenkins_setup.md]: You could reference our [verions.yaml](https://github.com/kata-containers/runtime/blob/master/versions.yaml) file and follow our 'minimum golang version' as recorded in there.

File: Jenkins_setup.md [hunk 11]
```diff
@@ -8,6 +8,10 @@
         * [Installing Plugins](#installing-plugins)
         * [Naming jobs](#naming-jobs)
         * [Setting up the scripts](#setting-up-the-scripts)
+    * [Setting up a bare metal Jenkins Slave Node](#setting-up-a-bare-metal-jenkins-slave-node)
+        * [Setting up a jenkins user](#setting-up-a-jenkins-use)
```
[the remark this answers is not in the document]
Comment (grahamwhaley) [reply Jenkins_setup.md]: ah, which might indicate a hand-crafted ToC update. @jodh-intel - do you have a recommended current tool to generate/update ToCs to ensure they are accurate and in the right format/indent level etc. I'm not sure if you just landed something or if that is in an in-flight PR. Might help @Pennyzct  make the new entries :-)

[the remark this answers is not in the document]
Comment (jodh-intel) [reply Jenkins_setup.md]: It's worth pointing out that when we get the static checks running for this repo, they will detect errors like the one @chavafg noticed (good spot btw! ;)

@grahamwhaley - yes, you can generate the TOC by running the following:

```sh
$ cd $GOPATH/src/github.com/kata-containers/tests/cmd/check-markdown
$ make
$ kata-check-markdown toc /path/to/your/README.md > /tmp/toc.md
```

Then, just paste `/tmp/toc.md` into the top of your `README.md`. The eventual plan is for that tool to both detect missing TOCs and have the option to rewrite the file with the TOC maybe.

File: Jenkins_setup.md (hunk 5 above)
Comment (Pennyzct) [reply Jenkins_setup.md]: super coooooool, real magic, thanks!!!!😍

File: Jenkins_setup.md (hunk 6 above)
Comment (Pennyzct) [reply Jenkins_setup.md]: super cool for poor-english me. ;)

File: Jenkins_setup.md (hunk 5 above)
Comment (jodh-intel) [reply Jenkins_setup.md]: I :heart: Linux/Unix :smile:

File: Jenkins_setup.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Pennyzct) [reply Jenkins_setup.md]: Emmmmmm, I might take the courage to ask what is ToC? 🤣

[the remark this answers is not in the document]
Comment (jodh-intel) [reply Jenkins_setup.md]: Sorry @Pennyzct - I should have been clearer. I was referring to a "Table of Contents" - the tool above will create the table of links for you, so you can add it into the top of the document itself :smile:

**Comment by Pennyzct** [comment]
Hi~ @grahamwhaley @jodh-intel @chavafg I have done all the refinement. ;) ptal. thanks again for the suggestion!!!!!

**Pull request merged**